--- warden.conf.orig	2016-07-19 05:02:47.141298919 +0000
+++ warden.conf	2016-07-19 05:03:30.838481468 +0000
@@ -65,12 +65,12 @@
 service.command.tt.heapsize.percent=2
 service.command.tt.heapsize.max=325
 service.command.tt.heapsize.min=64
-service.command.hbmaster.heapsize.percent=4
-service.command.hbmaster.heapsize.max=512
-service.command.hbmaster.heapsize.min=64
-service.command.hbregion.heapsize.percent=25
-service.command.hbregion.heapsize.max=4000
-service.command.hbregion.heapsize.min=1000
+service.command.hbmaster.heapsize.percent=25
+service.command.hbmaster.heapsize.max=5120
+service.command.hbmaster.heapsize.min=5120
+service.command.hbregion.heapsize.percent=50
+service.command.hbregion.heapsize.max=24576
+service.command.hbregion.heapsize.min=24576
 service.command.cldb.heapsize.percent=8
 service.command.cldb.heapsize.max=4000
 service.command.cldb.heapsize.min=256
@@ -95,7 +95,7 @@
 service.nice.value=-10
 zookeeper.servers=stl-colo-srv136.splicemachine.colo:5181
 nodes.mincount=1
-services.retries=3
+services.retries=1
 cldb.port=7222
 mfs.port=5660
 hbmaster.port=60000
