--- zoo.cfg.orig	2016-07-19 04:57:20.016946497 +0000
+++ zoo.cfg	2016-07-19 04:57:22.828958127 +0000
@@ -11,7 +11,7 @@
 # the port at which the clients will connect
 clientPort=5181
 # max number of client connections
-maxClientCnxns=100
+maxClientCnxns=0
 #autopurge interval - 24 hours
 autopurge.purgeInterval=24
 #superuser to allow zk nodes delete
@@ -26,3 +26,4 @@
 authProvider.1=org.apache.zookeeper.server.auth.SASLAuthenticationProvider
 # use maprserverticket not userticket for auth
 mapr.usemaprserverticket=true
+maxSessionTimeout=120000
