diff --git a/splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/operations/ExplainOperation.java b/splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/operations/ExplainOperation.java
index 5195ccd..50ebef0 100644
--- a/splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/operations/ExplainOperation.java
+++ b/splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/operations/ExplainOperation.java
@@ -147,7 +147,7 @@ public class ExplainOperation extends SpliceBaseOperation {
         String sql = activation.getPreparedStatement().getSource();
         Collection<QueryTreeNode> opPlanMap = m.get(sql);
         if(opPlanMap!=null){
-            explainStringIter = PlanPrinter.planToString(opPlanMap);
+            explainStringIter = PlanPrinter.planToIterator(opPlanMap);
         }else
             explainStringIter =Iterators.emptyIterator();
     }
